sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_data_lake_file_client.py

```diff
@@ -226,6 +229,110 @@ def get_file_properties(self, **kwargs):
         blob_properties = self._get_path_properties(**kwargs)
         return FileProperties._from_blob_properties(blob_properties)  # pylint: disable=protected-access
 
+    def _upload_options(  # pylint:disable=too-many-statements
+            self, data,  # type: Union[Iterable[AnyStr], IO[AnyStr]]
+            length=None,  # type: Optional[int]
+            **kwargs
+        ):
+        # type: (...) -> Dict[str, Any]
+
+        encoding = kwargs.pop('encoding', 'UTF-8')
+        if isinstance(data, six.text_type):
+            data = data.encode(encoding) # type: ignore
+        if length is None:
+            length = get_length(data)
+        if isinstance(data, bytes):
+            data = data[:length]
+
+        if isinstance(data, bytes):
+            stream = BytesIO(data)
+        elif hasattr(data, 'read'):
+            stream = data
+        elif hasattr(data, '__iter__'):
+            stream = IterStreamer(data, encoding=encoding)
+        else:
+            raise TypeError("Unsupported data type: {}".format(type(data)))
+
+        validate_content = kwargs.pop('validate_content', False)
+        content_settings = kwargs.pop('content_settings', None)
+        metadata = kwargs.pop('metadata', None)
+        max_concurrency = kwargs.pop('max_concurrency', 1)
+
+        kwargs['properties'] = add_metadata_headers(metadata)
+        kwargs['lease_access_conditions'] = get_access_conditions(kwargs.pop('lease', None))
+        kwargs['modified_access_conditions'] = get_mod_conditions(kwargs)
+
+        if content_settings:
+            kwargs['path_http_headers'] = get_path_http_headers(content_settings)
+
+        kwargs['stream'] = stream
+        kwargs['length'] = length
+        kwargs['validate_content'] = validate_content
+        kwargs['max_concurrency'] = max_concurrency
+        kwargs['client'] = self._client.path
+
+        return kwargs
+
+    def upload_data(self, data,  # type: Union[AnyStr, Iterable[AnyStr], IO[AnyStr]]
+                    length=None,  # type: Optional[int]
+                    overwrite=False,  # type: Optional[bool]
+                    **kwargs):
```

Comment:
I'm wondering if this should be called `upload_file`....
It seems that it cannot be used to append data to an existing file - is that correct?

Reply:
They can upload data to an empty file by specifying the etag and match condition to make sure they are flushing data to the existing empty file.
Otherwise things could become: 
`Customer1` created an empty file,
`Customer2` appended data and flushed.
`Customer1` still thinks the file is empty and do append and flush then the data will be wrong.

While it's still possible that `Customer2` appended data but didn't flush, in this case only lease is helpful.